test/upgrade/steps.go

```diff
@@ -63,17 +63,20 @@ func (se *suiteExecution) startContinualTests(num int) {
 					return
 				}
 				setup := operation.Setup()
+
+				logger, buffer := NewInMemoryLoggerBuffer()
 				t.Run("Setup"+operation.Name(), func(t *testing.T) {
-					setup(Context{T: t, Log: l})
+					setup(Context{T: t, Log: logger})
 				})
 				handler := operation.Handler()
 				go func() {
-					bc := BackgroundContext{Log: l, Stop: operation.stop}
+					bc := BackgroundContext{Log: logger, LogBuffer: buffer, Stop: operation.stop}
 					handler(bc)
 				}()
 
 				se.failed = se.failed || t.Failed()
 				if se.failed {
+					t.Log(buffer)
```

Comment:
Each line should be logged by it self, instead of whole buffer in one go.

Also, I think we should log no mater that test fail or not. That's the job of `go test -v` to display those logs or not.

Reply:
I agree with therest, but why log every line separately?

Reply:
As we are not logging them when they occur, we should save emit time. Then log those log entries with that emit time. That's to ensure we can see how much time given operations take.

Reply:
Sorry about this. You are still using zap logger, and he adds the times.

So, it should be okay to print all of them at once.